tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -329,75 +394,192 @@ def get_def_segment_at_position(
+        workspace_config = self.get_config_for_uri(uri)
+
+        if path in self.contexts:
+            return self.contexts[path]
+
+        context = ContextContainer(path, workspace_config)
+        for file in context.file_paths:
+            self.contexts[file] = context
+        self.contexts[path] = context
+
+        return context
+
+    def get_file_info(
+        self, uri: str, do_update: bool = False
     ) -> Tuple[FileInfo, List[Any]]:
         """
-        The LSP driver maintains one Chapel context per-file. If there is no context, this function creates a context. If `do_update` is set, this function assumes the file content has change and advances revisions.
+        The language server maintains a FileInfo object for file. The FileInfo
+        contains precomputed information (binary-search-ready tables for
+        finding an element under a cursor).
+
+        This method retrieves the FileInfo object for a particular URI,
+        creating one if it doesn't exist. If do_update is set to True,
+        then the FileInfo's index is reuilt even if it has already been
+        computed. This is useful if the underlying file has changed.
         """
 
         errors = []
 
-        if uri in contexts:
-            file_info = contexts[uri]
+        if uri in self.file_infos:
+            file_info = self.file_infos[uri]
             if do_update:
-                file_info.context.advance_to_next_revision(False)
-                with file_info.context.track_errors() as errors:
-                    file_info.rebuild_index()
+                errors = file_info.context.advance()
         else:
-            context = chapel.core.Context()
-            with context.track_errors() as errors:
-                file_info = FileInfo(uri, context)
-            contexts[uri] = file_info
+            file_info, errors = self.get_context(uri).new_file_info(uri)
+            self.file_infos[uri] = file_info
 
         return (file_info, errors)
 
-    def build_diagnostics(uri: str) -> List[Diagnostic]:
+    def build_diagnostics(self, uri: str) -> List[Diagnostic]:
         """
         Parse a file at a particular URI, capture the errors, and return then
         as a list of LSP Diagnostics.
         """
 
-        fi, errors = get_context(uri, do_update=True)
+        _, errors = self.get_file_info(uri, do_update=True)
 
         diagnostics = [error_to_diagnostic(e) for e in errors]
         return diagnostics
 
+    def register_workspace(self, uri: str):
+        path = os.path.join(uri[len("file://") :], ".cls-info.json")
+        config = WorkspaceConfig.from_file(self, path)
+        if config:
+            self.configurations[uri] = config
+
+    def unregister_workspace(self, uri: str):
+        if uri in self.configurations:
+            del self.configurations[uri]
+
+
+def run_lsp():
+    """
+    Start a language server on the standard input/output
+    """
+    server = ChapelLanguageServer()
```

Comment:
Thanks for making this change!

Reply:
I've been dreaming about per-server state since the beginning. So this was an obvious move. Thanks for bringing extending `LanguageServer` to my attention